### ISSUE TITLE
DeepSpeed launch jobs with Kubernetes

### ISSUE THREAD

**Issue by g-karthik** [microsoft/DeepSpeed#274]
I'm hoping to launch DeepSpeed-based multi-node training jobs with Kubernetes, and for this purpose, I'm currently working on writing YAML files for my jobs and building a custom Docker image with all needed third-party dependencies (including DeepSpeed).

I see in the [resource-configuration-multi-node](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node) section of the docs that DeepSpeed typically requires a hostfile for multi-node training, but that kinda defeats the purpose of using Kubernetes.

What is a good way to enable DeepSpeed-based multi-node training with Kubernetes?

I see in the [mpi-compatibility](https://www.deepspeed.ai/getting-started/#mpi-compatibility) section that DeepSpeed is compatible with mpirun.

I quote the docs:

DeepSpeed will then use mpi4py to discover the MPI environment (e.g., rank, world size) and properly initialize torch distributed for training. In this case you will explicitly invoke python to launch your model script instead of using the deepspeed launcher, here is an example:

`mpirun <mpi-args> python \
	<client_entry.py> <client args> \
	--deepspeed_mpi --deepspeed --deepspeed_config ds_config.json`

Is mpirun the solution for training with Kubernetes? What kind of args need to be passed as `<mpi-args>` above?

Do you have any sample multi-node training job YAML files and Dockerfiles that you could share? Could you summarize a checklist of things to do to the training code, environment variables to set, etc. (I may be missing something here) to make it ready for Kubernetes-based multi-node training?

Thanks!

**Comment by jeffra** [comment]
We are actually using Kubernetes in some of our deployments of DeepSpeed (via https://github.com/microsoft/dlworkspace but in this setup DLWorkspace creates the yaml files on the fly for us per job). Can you tell me a bit more about your setup? Perhaps share a sample yaml you are using?

Area you already able to use `mpirun` for different applications? If so then yes you can launch DeepSpeed with `mpirun` via the example you link above. However commonly mpi will require a mpi-style hostfile that is the same format as what DeepSpeed requires. The `<mpi-args>` are whatever mpi arguments you might need in your own setup to just run mpi and expose the relevant interfaces for your hardware (e.g., IB related arguments).

Essentially the DeepSpeed launcher just needs to know the hostnames in your job and have passwordless-ssh setup between nodes so that they can communicate. The hostfile described in the [resource-configuration-multi-node](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node) section lists the available worker nodes in the job so that DeepSpeed knows how to communicate with all the nodes.

**Comment by Rachnas** [comment]
@g-karthik, @jeffra, I am in the same boat, trying to run multi node training with Kubernetes using ds_zero2_pretrain_gpt2_model_parallel.sh script.  I have some doubts about hostfile, given 2 nodes, 1 GPU per node. Do I need to mention both the machine names in hostfile? A sample hostfile will be helpful. 
I created host file as described in the resource-configuration-multi-node section, but ssh fails for me. I do not have much control on the environment.

**Comment by jeffra** [comment]
@Rachnas the hostfile should be something like the example in [resource-configuration-multi-node](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node)

```
worker-1 slots=4
worker-2 slots=4
```

Where in the above hostfile `worker-1` should be ssh-able with that name either by hostname or [ssh-config alias](https://www.ostechnix.com/how-to-create-ssh-alias-in-linux/) and you should be able to run this command without a password prompt: `ssh worker-1 date` and it should return the date on the remote node.

**Comment by Rachnas** [comment]
Thanks @jeffra I did the same, mention IP addresses in place of worker-1 and worker-2. Is it correct ?

**Comment by jeffra** [comment]
@Rachnas IP addresses should work here also, I don't see why not. For context, the deepspeed launcher uses [pdsh](https://linux.die.net/man/1/pdsh) in the backend to launch processes across nodes so as long as the nodes are ssh-able like i describe above then you should be good to go.

**Comment by Rachnas** [comment]
@jeffra, yes looks like IP address worked. But it fails for me at ssh authentication 
connect to host <> port <>: Connection refused

**Comment by jeffra** [comment]
@Rachnas are you using our DeepSpeed docker image or your own? Our image sets the ssh port to 2222 by default: https://github.com/microsoft/DeepSpeed/blob/master/Dockerfile#L87-L89 

Can you share the output of these two commands on one of the worker nodes?
1) `grep "Port " /etc/ssh/sshd_config`
2) `service ssh status`

**Comment by Rachnas** [comment]
@jeffra Thanks for pointing it out. I am not using DeepSpeed docker image.  Command outputs are:
* sshd is running
#Port 22

**Comment by jeffra** [comment]
Interesting. Can you successfully `ssh localhost date` (or replace localhost with the ip address of the local node you are trying this on)? I am just concerned that your ssh server isn't setup properly.

If it's a password issue you'll want to make sure you have setup your key properly in ~/.ssh/. I would recommend some reading on posts similar to this: https://askubuntu.com/questions/46930/how-can-i-set-up-password-less-ssh-login

**Comment by Rachnas** [comment]
@jeffra thank you so much. I think my password issues is because of kubernetes setup I am working in (jumpbox, PODs etc). Not possible to give much details :) . I will try to solve this and update outcome here.

**Comment by jeffra** [comment]
Sounds good, this at least points you to what I believe is the core issue. If you can solve the passwordless SSH issue then you should be good to go to run DeepSpeed in your environment.

**Comment by Rachnas** [comment]
@jeffra I am hoping the same. This is such a great learning, many thanks to you and your team for DeepSpeed.

**Comment by g-karthik** [comment]
@jeffra thanks for your reply! I managed to get this working with mpirun and deepspeed!

On a somewhat unrelated note, could you please elaborate on the need for each of these dependencies in deepspeed's [Dockerfile](https://github.com/microsoft/DeepSpeed/blob/master/Dockerfile#L58L89)?

**Comment by jeffra** [comment]
@g-karthik Excellent news, glad you got things running. Do you have any suggestions on how we can improve our documentation so that others don't run into the issues you had to go through to get deepspeed/mpirun to work in your environment?

More details on the components that you highlight in our dockerfile: 
* Mellanox OFED relate to InfiniBand (IB) drivers we need for our Azure and DGX-2 clusters, they might not be needed for clusters that don't have IB.
* nv_peer_mem also relates IB. Specifically [this package](https://github.com/Mellanox/nv_peer_memory) allows support for GPUDirect RDMA support ([some further reading](https://www.mellanox.com/sites/default/files/related-docs/prod_software/PB_GPUDirect_RDMA.PDF) if you're curious).
* The block relates to setting the ssh daemon config file to use a specific port number (in this case 2222). This is mostly required for our [Azure VM example](https://github.com/microsoft/DeepSpeed/tree/master/azure) to work since the default ssh port of 22 connects to the host OS and this is the ssh port for the container itself. In our internal environment however our k8s setup will give us a ssh port for each node in a given job. Was this part of your original issue perhaps?

**Comment by g-karthik** [comment]
@jeffra Thanks for the info! Yes I think I will be needing the Mellanox plugin and RDMA support as well to increase communication bandwidth for my setup, which is based on AWS EKS. I didn't do any port stuff in the Dockerfile, I just installed OpenMPI and OpenSSH.

In terms of improving documentation, I think perhaps you could mention that an explicit hostfile isn't necessarily needed (depending on the environment). It would be even better if you could provide a detailed example (with YAMLs, etc.) of enabling multi-node training with all major cloud providers' services (AWS, Azure, GCP) so that more people are able to use DeepSpeed across the board.

**Comment by yochze** [comment]
FYI - we've managed to use DeepSpeed on Kubernetes with OpenMPI and the [MPI Operator](https://github.com/kubeflow/mpi-operator/) using DeepSpeed's Dockerfile with added OpenMPI and SSH configs: https://gist.github.com/yochze/e6954524da6bc0a080c43015dead4903

**Comment by Gforky** [comment]
Can anybody share a Kubernetes's YAML file for multi-node training using PDSH runner? Since after I read about the implementation of  PDSH runner, I noticed that there should be one node to run the pdsh script, so this node should be the initial container in a, for example, PyTorch job cluster? Please correct me if I'm wrong.
Is there any plan to provide a native runner, that each worker in the cluster run the same script, same we manually run the multi-node PyTorch training using torch.distributed.launcher?
Thank you guys in advance.

**Comment by edwardzjl** [comment]
> @Rachnas are you using our DeepSpeed docker image or your own? Our image sets the ssh port to 2222 by default: https://github.com/microsoft/DeepSpeed/blob/master/Dockerfile#L87-L89
> 
> Can you share the output of these two commands on one of the worker nodes?
> 
> 1. `grep "Port " /etc/ssh/sshd_config`
> 2. `service ssh status`

The Dockerfile has not been updated for a while, according to the file content on master branch, the ssh port was still 22.

There are 2 sections modifying ssh config. The first section uncomment the `#Port 22` to `Port 22`:

https://github.com/microsoft/DeepSpeed/blob/17fa0876adcc8fa2acb384727bdc49d7434f1f99/docker/Dockerfile#L30-L36 

And the second section tries to replace the origin `#Port 22` to `Port 2222` which is not working:

https://github.com/microsoft/DeepSpeed/blob/17fa0876adcc8fa2acb384727bdc49d7434f1f99/docker/Dockerfile#L150-L155 

Exec into the container the port was still 22:

```
...
Include /etc/ssh/sshd_config.d/*.conf

Port 22
#AddressFamily any
#ListenAddress 0.0.0.0
#ListenAddress ::
...
```

Maybe the Dockerfile need to be updated?

Sorry if I poluted this issue, or should I open a new one?

**Comment by h6mx** [comment]
so have somebody use deepspeed successfully in k8s ? i have trouble making different pod communicate with each other, how to write the hostfile in k8s ? Thank u so much !

**Comment by yiliu30** [comment]
> so have somebody use deepspeed successfully in k8s ? i have trouble making different pod communicate with each other, how to write the hostfile in k8s ? Thank u so much !

As yochze mentioned [above](https://github.com/microsoft/DeepSpeed/issues/274#issuecomment-664180458), there is no need to write a host file for using Deepspeed regarding k8s. The [MPI Operator](https://github.com/kubeflow/mpi-operator/) will manage the nodes(pods). You can treat the Deepspeed task as a normal mpi task with k8s. 
What you need to do is specify a `Launcher` and `Worker`` along with a specified quantity of replicas.

Here's an example of how to launch an MPI task with 2 processes with k8s.
https://github.com/kubeflow/mpi-operator/blob/master/examples/v2beta1/horovod/tensorflow-mnist.yaml

**Comment by h6mx** [comment]
> > so have somebody use deepspeed successfully in k8s ? i have trouble making different pod communicate with each other, how to write the hostfile in k8s ? Thank u so much !
> 
> As yochze mentioned [above](https://github.com/microsoft/DeepSpeed/issues/274#issuecomment-664180458), there is no need to write a host file for using Deepspeed regarding k8s. The [MPI Operator](https://github.com/kubeflow/mpi-operator/) will manage the nodes(pods). You can treat the Deepspeed task as a normal mpi task with k8s. What you need to do is specify a `Launcher` and `Worker`` along with a specified quantity of replicas.
> 
> Here's an example of how to launch an MPI task with 2 processes with k8s. https://github.com/kubeflow/mpi-operator/blob/master/examples/v2beta1/horovod/tensorflow-mnist.yaml

oh, Thanks !  I will try this! I also wanna know how the mpi operator work and how mpi operator manage deepspeed task . what should i do ? thanks for your time

**Comment by ekkkkki** [comment]
In the image built from this [dockerfile](https://gist.github.com/yochze/e6954524da6bc0a080c43015dead4903), I started Deepspeed under MPI like `mpirun -{mpiconfig} deepspeed --num_pods=2 script.py --{pyconfig}`(I requested 2 nodes in k8s)
however I was told `Unable to find hostfile, will proceed with training with local resources only.`

appreciate for any suggestion.

**Comment by yiliu30** [comment]
> In the image built from this [dockerfile](https://gist.github.com/yochze/e6954524da6bc0a080c43015dead4903)(with `mpi4py`installed), I started Deepspeed under MPI like `mpirun -{mpiconfig} deepspeed --num_pods=2 script.py --{pyconfig}` however I was told `Unable to find hostfile, will proceed with training with local resources only.`
> 
> Thanks a lot if you have any suggestion.

Did you install the [Kubeflow Training Operator](https://github.com/kubeflow/training-operator#stable-release) ?

**Comment by Qingyuncookie** [comment]
```
sudo apt update
sudo apt-get install libopenmpi-dev
pip install mpi4py
```

**Comment by abdulvirta** [comment]
> @Rachnas are you using our DeepSpeed docker image or your own? Our image sets the ssh port to 2222 by default: https://github.com/microsoft/DeepSpeed/blob/master/Dockerfile#L87-L89
> 
> Can you share the output of these two commands on one of the worker nodes?
> 
> 1. `grep "Port " /etc/ssh/sshd_config`
> 2. `service ssh status`

@jeffra is there a release of this image somewhere on dockerhub? I couldn't find it. should we try to build it ourselves? In general what is the recommended base-image to use with [DeepSpeedChat](https://github.com/microsoft/DeepSpeed/tree/master/blogs/deepspeed-chat)? I'm also having a hard time getting ssh to work for multi-node training in K8s
